### ISSUE TITLE
Suggestions to improve the code completion widget

### ISSUE THREAD

**Issue by ghost** [spyder-ide/spyder#2980]
Automatic On context information + persistence of this information in the right panel.
You mention the Goto Sublime, it's better to have persistently on the right side panel+auto switch

Example when typing:
Currently we have:
0)   np.histog     + Select  histogram2d  in pop list +  Enter    :       histogram2d   is written on Editor

Other function very important:
1)   np.histog     + Select  histogram2d in pop list +  space
       --> Switch automatically the Right Side panel to Object inspector
            +  show the following info about   np.histogram2d 
            +  Pop list is maintained
            BUT "histogram2d  " is NOT written in the Editor
 ---> Allow to investigate function by using pop up  by staying in the editor

2)  np.histog     + Select  histogram2d in pop list  +  Alt
      -->  Write down in editor:    histogram2d(x, y, bins=10, range=None, normed=False, weights=None)
              + Switch automatically the Right Side panel to Object inspector   
              +  show the following info about   np.histogram2d   in that panel
(For speed purpose, Text info with no formatting is sufficient)

3) np.histog     + Select  histogram2d  +  AltGr 
      --> Launch "Chrome/Default Browser"  + Google.com + search:     python "import numpy"   histogram2d       ("import package_name") + function/attribut

4) np.histog     + Select  histogram2d  +  Ctrl  
     --> Right Side panel (Object Inspector or other), show Source Code file/ Source Code of numpy histogram2d
            + FolderName+Filename
           (it looks like your Goto function)

A lot of time is spent to search the correct syntax/function to use/function available/code....
automation of this  will save a lot of time and the features outstanding compare to other editors.

**Comment by arita291** [comment]
Hello, 
 Thank you for putting in 3.1 release.
 Grabbing information while typing --> It would make Spyder the Top One IDE,
 because a lot of time is spent how to put together exising code.
 
 if you can indicate the location of the code for auto-completion in GitHub,
 I woud love to have a look and think about simple ideas to make Spyder a top one editor.

  Sypder is already pretty good one, congratulation.

**Comment by arita37** [comment]
Hello,

Wondering if those features will be on the next version of Spyder, it would be very useful ones !

**Comment by Nodd** [comment]
It's planned for 3.1 for now, so it won't be in the next release (3.0).

I think some of the features you presented would fit better in the object inspector rather than the code completion widget (points 3 and 4 for example).

Also having so many modifier keys in a single popup looks very confusing. I'd never remember all of them.

For 1 the problem is that if you type some code it will be very common to type a variable name followed by space. In this case what the user wants is to continue typing without being bothered by the popup menu.

So maybe these ideas are great but they should be refined before being implemented.

**Comment by arita37** [comment]
Hello,
Thanks for your comments.

Main idea is when we type code, especially for new library..
we often look for information about the nature of the function:
      "what it does: input, output,......  and how to use it"

As soon as, we nee to take the mouse ---> we lose time,
also we need often need to copy/paste into Google / Documentation...

Current Object inspector is **very slow** and often not responding....
Maybe, Text based (raw) would accelerate the Object inspector.


1) np.histog + Select histogram2d in pop list + space
     You need to select 1st the histogram  ONLY selection if you type "space", it can be done.

2) Please have a look at your keyboard:
   Space, Alt, AltGr, Ctrl   ---> all keys can be input but one finger  when typing
    
3) When you have a new library (not numpy..., a new one), how many times did you use your mouse
to look in google for information on the library syntax, examples.....
or even look at the source code for "semi-debugging"....

**Comment by Nodd** [comment]
You can already use raw text for the object inpector (menu in the topright corner). In my experience the delay seems related to finding the documentation for the object rather than displaying it, because it feels like the same delay for the autocompletion popup to appear.

1) An entry is already selected when the tooltip appears, so that pressing enter validates the suggestion.
2) My problem is not typing the shorcuts, it's remembering which is which. Also it's not discoverable at all.
3) I have no problem with facilitating the access to documentation, be it inline, google, source code... It just has to be done in a smart way. I already use the mouse quite a lot, and for me added links or buttons in the outline explorer is easier to use and remember than packing everything in modifiers from the popup menu. I think other editors don't do it like this for a reason!

**Comment by arita37** [comment]
0. Object inspector is very slow and often breaks on many packages...
   (if you can run the benchmark and send it, it would be nice)

   Auto Completion is fast.

1) This is not about to validate an entry...
     This is to get the information about a method suggests by auto-completion:
        Input arguments, method output, method Doc, method details
     without using the mouse, and going to google and other broken parts like object inspector


2) You can add the shorcut info in Spyder doc/ shortcut info.
     Of course, if you dont put any info ---> "it is not discoverable at all"....
    (your comment is very strange)

3) Other editor does it: IntelliJ and Pycharm has better auto-completion info.

4) When you have a new package liek PyTorch:
     How many times do you have to go to Google / Doc just to get the meaning of one method,
     output details, type,..... ?

5) Just try yourself with a new package, how long do you need to get the info, just to lake run the method....

7) Am open for discussion, if you have better ideas to speed up the typing without using mouse.

**Comment by arita37** [comment]
Attached IntelliJ editor for C++ or Java 

![image](https://cloud.githubusercontent.com/assets/18707623/15240191/d67a5996-1922-11e6-92af-b68e17a2c4d1.png)

**Comment by Nodd** [comment]
It's a bit hard to follow the discussion because there are many different topics. I'll create separate issues.

If the object inspector breaks on some packages it's a bug that should be fixed. Feel free to open new issues or pull requests!
Did you try the text mode to see if performance improves ?

1) I know it's not about validating an entry, but adding functionality should not prevent from doing normal actions like typing code or validating entries. Binding an action on space does interfere with typing.

2) By discoverability, I mean not buried in a paragraph somewhere in the documentation. If it's in a menu somewhere (with a sortcut), there is much more chance that a user will see that the functionality is available.
If it's a modifier in the completion widget, it won't appear in the shortcut options, only actual shortcuts are there.

Adding the signature in the completion widget would be a very useful addition.

**Comment by goanpeca** [comment]
Agree with @Nodd, I think some suggestions are valid, some not but we definitely need to split them in separate more granular issues. 

Could you do it @Nodd? Add as many descriptive labels as you can :-)

Cheers

**Comment by arita37** [comment]
I think there is a misunderstanding :
"Binding an action on space does interfere with typing."

This does not interfere in that way:

Start typing:    np.hist
Widget appears with the list
2 choices:   continue to type  (widget is auto-updated), things as usual.
                 **OR start using the Up / Down key to select/navigatge the choices** 

IF ONLY we have used the Up/Down keys (=we are looking for some methods),
AND If we type a key when the method is highlighted in the widget
   THEN proceed to the extra functions (described above).

1) Up or Down Keys Must be pressed and method highlighted
    AND Special key must be type to get the extra function.

It does not interfere with typing since Up/Down key must be pressed.

**Comment by Nodd** [comment]
@arita37 The problem is with space with no modifier. An entry is already automatically selected (so that enter validates the entry), so the up/down arrow doesn't change anything here.

**Comment by Nodd** [comment]
I created separate issues, I think I forgot nothing. Feel free to comment on them or create others for other subjects.
I'll close this one so that the conversation can continue on each specific issue.
